<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html><head>














	<meta http-equiv="CONTENT-TYPE" content="text/html; charset=utf-8"><title>SigmoID Help</title>
	
	<meta name="GENERATOR" content="NeoOffice 3.4.1  (Unix)">
	<meta name="CREATED" content="0;0">
	<meta name="CHANGED" content="20150928;21232700"></head><body dir="ltr" lang="en-GB">
<p><br></p>
<ul id="mozToc"><!--mozToc h1 1 h2 2 h3 3 h4 4 h5 5 h6 6--><li><a href="#mozTocId526876">SigmoID overview</a></li><li><a href="#mozTocId182858">Installation</a></li><li><a href="#mozTocId233463">Supported file formats</a></li><li><a href="#mozTocId629072">Interface</a><ul><li><a href="#mozTocId49640">Windows</a><ul><li><a href="#mozTocId965351">Main Window</a></li><li><a href="#mozTocId23083">Genome Browser Window</a></li><li><a href="#mozTocId508441">Database Windows</a></li><li><a href="#mozTocId297884">Profile Wizard Window</a></li><li><a href="#mozTocId65146">Command Configuration Windows</a></li><li><a href="#mozTocId686762">Web Browser and Help Windows </a></li><li><a href="#mozTocId522291">SigmoID Preferences Window</a></li></ul></li><li><a href="#mozTocId385906">Menu Reference
</a><ul><li><a href="#mozTocId324242">File Menu</a></li><li><a href="#mozTocId121686">Edit Menu</a></li><li><a href="#mozTocId916489">View Menu</a></li><li><a href="#mozTocId494918">Profile Menu</a></li><li><a href="#mozTocId972700">Regulon menu</a></li><li><a href="#mozTocId693367">Genome Menu</a></li><li><a href="#mozTocId600980">
Window Menu</a></li><li><a href="#mozTocId2010">
Help Menu</a></li></ul></li></ul></li><li><a href="#mozTocId708060">Python Scripts</a><ul><li><a href="#mozTocId227412">HmmGen.py</a></li><li><a href="#mozTocId61963">MastGen.py</a></li><li><a href="#mozTocId891699">TermGen.py</a></li><li><a href="#mozTocId793586">ptt_converter.py</a></li><li><a href="#mozTocId218330">OperOn.py</a></li><li><a href="#mozTocId385189">gbk2tbl.py</a></li></ul></li></ul>


<br>

<h1><a class="mozTocH1" name="mozTocId526876"></a>SigmoID overview</h1>&nbsp;&nbsp;&nbsp; SigmoID front end is a GUI application written in <a href="http://xojo.com">Xojo</a>
which gives it the expected look and usability on all three supported
platforms (OS X, Linux and Windows). The GUI provides the interface for
selected programmes from the <a href="http://hmmer.org">HMMER</a>, <a href="http://meme-suite.org/">MEME Suite</a> and <a href="http://transterm.cbcb.umd.edu">TransTerm HP</a>, which are responsible for the actual searches. Sequence logos are calculated by <a href="http://weblogo.threeplusone.com">WebLogo</a>
which is written in Python, hence Python (version 2.7.x) and BioPython (version 1.64
and above) are required. Processing of nhmmer, mast and TransTerm HP
output, sequence format conversions and adding regulatory sites to
genome annotation are implemented as separate <a href="#Python_Scripts">Python scripts</a>. The
scripts are called from the GUI, but could easily be used separately
and integrated in an annotation pipeline if desired. Detailed
installation instructions are provided within distributions for each
platform. The <a href="https://github.com/nikolaichik/SigmoID">source code</a> for the whole SigmoID application is available with GPL 2.0 license. <br>
<br>&nbsp;&nbsp;&nbsp; SigmoID allows to:<br>&nbsp;&nbsp;&nbsp; - get binding site data from specialised databases;<br>&nbsp;&nbsp;&nbsp; - visualise binding site alignments with sequence logos;<br>&nbsp;&nbsp;&nbsp; - extend, shorten and mask alignments;<br>&nbsp;&nbsp;&nbsp; - create optimised hmm profiles from alignments;<br>&nbsp;&nbsp;&nbsp; - search bacterial genomes with calibrated (and uncalibrated) hmm profiles;<br>&nbsp;&nbsp;&nbsp; - add annotation of promoters and transcription factor binding sites to GenBank-formatted genome files;<br>&nbsp;&nbsp;&nbsp; - edit genome annotation.<br>
<br>
This version includes 43 calibrated profiles (for 5 sigma factors and 36 TFs) optimised for enterobacterial phytopathogens <span style="font-style: italic;">Pectobacterium</span> spp. and <span style="font-style: italic;">Dickeya dadantii</span>. Efficiency of these profiles may decrease for other bacteria. <br>
<br>
The search for binding sites is done by nhmmer which is expected to be
installed in the default location. WebLogo installation is required for
new alignments, otherwise logos won't be displayed, but the rest of
SigmoID should work. <br>
<br>
Adding annotations to GenBank files is done by the HmmGen.py script
which could be used on its own. BioPython (version 1.64 and up) is
required. <br>
<br>
At the moment SigmoID is known to work on OS X (10.8-10.11), Windows (Vista, 7
and 8), Ubuntu (12.04 and 14.04). It may also work on other Linux
distributions, provided the required libraries are installed. <br>
<br>
<h1><a class="mozTocH1" name="mozTocId182858"></a>Installation</h1>&nbsp;&nbsp;&nbsp;
SigmoID relies heavily on Python and will have severely limited
functionality without it. The various python scripts included with
SigmoID require python version 2.7 and will not work with version 3.
Linux and OS X systems should have python installed, while Windows
users can download it from <a href="https://www.python.org/downloads/windows/">python.org</a>.
Please note that on Windows you need to modify system PATH environment
variable to include path to python. The easiest way to do it is to
select the "Add python.exe to path" option in the installer (it's not
selected by default!). You may check that python is installed correctly
by typing 'python' at the command prompt, which should launch the
python interpreter.<br>
&nbsp;&nbsp;&nbsp; Biopython (v. 1.64+) should be installed on top of
python 2.7. You can download the distribution and read the installation
details for your system at <a href="http://biopython.org/wiki/Download">biopython.org</a>. Please make sure you download and install the correct version for you system and python version.&nbsp;<br>
&nbsp;&nbsp;&nbsp; Depending on your system set-up, MEME
may require additional python and perl modules. Please check the log
pane of SigmoID's window for error messages and install missing modules
if necessary. <br>
&nbsp;&nbsp;&nbsp; There are two Linux distributions of SigmoID, for
32-bit and 64-bit systems. Both may require installation of additional
libraries. The 32-bit version depends on WebKit version 1 for
displaying help and database search results. WebKit1 is actually
included with the 32-bit Linux distribution; please see the SigmoID.sh
file for the correct command to launch SigmoID with included WebKit
libraries.<br>
<br>
<h1><a class="mozTocH1" name="mozTocId233463"></a>Supported file formats<br>
</h1>
&nbsp;&nbsp;&nbsp; SigmoID can open two types of sequence files: with genome sequences and&nbsp; TFBS/promoter sequences.<br>
&nbsp;&nbsp;&nbsp; SigmoID expects annotated genome sequence files to
be in GenBank format. The current version of genome browser can
properly open only files with a single accession per file.&nbsp; You
can still perform a nhmmer (but not MAST) search of the genome split
into several separate accessions, but please don't try to open this
type of GenBank files in genome browser.<br>
&nbsp;&nbsp;&nbsp; SigmoID can also work with unannotated&nbsp; genome
sequences in fasta format.&nbsp; An unannotated genome in fasta format
can't be opened directly in&nbsp; genome browser, but can be selected
as a target for nhmmer/MAST search.<br>
&nbsp;&nbsp;&nbsp; TFBS/promoter sequences can be in either fasta or
special SigmoID profile format (with .sig extension). TFBS/promoter
sequences in fasta files should be aligned and of the same length. If
facing varying lengths sequences, SigmoID displays error meaasage and
doesn't show sequence logo; the sequences, however, are loaded (for
viewing purposes and/or for aligning them properly with MEME).<br>
&nbsp;&nbsp;&nbsp; <span style="color: rgb(0, 102, 0); font-weight: bold;">SigmoID profile format</span>
files (.sig extension) are virtual folders containing several separate
files: the actual binding site sequences in fasta format, calibrated
hmm and MEME profiles, as well as two text files with profile
description and search engine/postprocessing options. The contents of
the files within the .sig virtual folder can be viewed in the <a href="#mozTocId965351">main SigmoID window</a> via commands from the <a href="#mozTocId916489">View menu</a>.&nbsp;
SigmoID has two hidden menu commands for converting a real folder to
the .sig file and vice versa; holding down the&nbsp; 'Alt' button while
selecting the File menu reveals them. <br>
&nbsp;&nbsp;&nbsp; Please note that due to current Xojo limitation
SigmoID can't handle virtual folders properly on 64-bit machines,
therefore 64-bit Linux version converts all .sig files to real folders.
These can only be accessed via toolbar of the main window and can't be
opened via any menu command. Therefore, on 64-bit Linux machine, please
move to the current profile folde any .sig folders you want to open and
use the leftmost toolbar button of the main window to open them.<br>
&nbsp;&nbsp;&nbsp; SigmoID can save genome files in standard GenBank
format, export unannotated genome sequence in fasta format and export
feature table in Sequin table format required by the NCBI tbl2asn
program.&nbsp; The appropriate commands are located in the <a href="#mozTocId324242">File menu</a>.  <br>

<h1><a class="mozTocH1" name="mozTocId629072"></a>Interface</h1>

<h2><a class="mozTocH2" name="mozTocId49640"></a>Windows</h2>
<h3><a class="mozTocH3" name="mozTocId965351"></a>Main Window</h3>
&nbsp;&nbsp;&nbsp; The main window opens at SigmoID launch and is split into two major interface elements: the topmost <span style="font-weight: bold; color: rgb(0, 102, 0);">Viewer</span> and the <span style="color: rgb(0, 102, 0); font-weight: bold;"><a name="LogPane"></a>Log</span> pane behind it. The <span style="color: rgb(0, 102, 0); font-weight: bold;">Log</span>
displays informative messages (including errors) and also shows textual
output from some of the included command line programmes and python
scripts. The <a name="Viewer"></a><span style="font-weight: bold; color: rgb(0, 102, 0);">Viewer</span>
is hidden at SigmoID launch and opens once binding site data is loaded.
It displays the sequence logo by default, but can be switched to
display other info for the loaded data via the <a href="#mozTocId916489">View menu</a>.
The info that can be shown by the Viewer depends on the type of the
file opened: only the logo and the sequences could be shown for fasta
files, while all options are available for .sig files. The <span style="color: rgb(0, 102, 0); font-weight: bold;">Viewer</span>
can be used to edit binding site sequences which has two consequences:
sequence logo is recalculated and all profile settings are discarded
since they become invalid (this is also reflected in the <a href="#mozTocId297884">Profile Wizard</a>
window). Please note that other types of information (settings,
description, hmm profile, etc.) should not be edited here as this will
have no effect. Only the changes made via the <a href="index.html#mozTocId297884">Profile Wizard</a> window can be saved (in a .sig file) and reused.<br>
&nbsp;&nbsp;&nbsp; The sequence logo displayed by the <span style="font-weight: bold; color: rgb(0, 102, 0);">Viewer</span>
is interactive and allows part(s) of the alignment to be selected. A
single area can be selected by dragging a mouse across the logo;
additional area(s) could then be added by pressing the "Shift" button
and dragging again. Such a selection can have two possible uses. First,
you can save the sequences corresponding to the selected area of the
logo in a new fasta file (via the <a href="#Save_Profile_Selection">Save Profile Selection...</a>
command from File menu). Second, you can launch the nhmmer search with
selected parts of the alignment masked. The masking happens by default
if you initiate a search when a part of the alignment is selected. This
works by invoking the alimask programme from the HMMER package to
produce a masked hmm profile which is then used to search the target
genome. You can set masking options in the nhmmer configuration window.
Please refer to HMMER User Guide for the details. <br>
&nbsp;&nbsp;&nbsp; The toolbar contains the buttons for few of the most used functions. The leftmost "<span style="font-weight: bold; color: rgb(0, 102, 0);">Load Alignment</span>"
button allows to open binding site data from either fasta or .sig
files. The .sig files provided with SigmoID could be chosen from the
drop down menu. User files can be opened by just pressing this button
(on Linux or Windows) or choosing More... at the very bottom of the
drop down list (OS X).&nbsp; <br>
&nbsp;&nbsp;&nbsp; The next "<span style="font-weight: bold; color: rgb(0, 102, 0);">Search</span>" button launches nhmmer search with the currently loaded profile. The raw search results appear in the <span style="font-weight: bold; color: rgb(0, 102, 0);">Log</span> pane. In case a .sig file is opened the post processing python script is launched&nbsp; by default and the <a href="#mozTocId23083">Genome Browser window</a>
is opened to display search results. For non-calibrated profiles
(opened from fasta files) the post processing script has to be launched
separately by pressing the third "<span style="font-weight: bold; color: rgb(0, 102, 0);">PostProcess</span>" toolbar button.<br>
&nbsp;&nbsp;&nbsp; The fourth "<span style="font-weight: bold; color: rgb(0, 102, 0);">Terminators</span>"
toolbar button allows to search for terminators. This function uses
TransTerm HP, performs the necessary format conversions and adds the
terminators to genome annotation. As TransTerm takes some time to run,
the results may take couple of minutes to appear, depending on
available processing power. <br>
&nbsp;&nbsp;&nbsp; The fifth "<span style="font-weight: bold; color: rgb(0, 102, 0);">Palindromise</span>"
button does a simple thing - it reverse complement sequences of the
currently loaded binding sites, adds them to the currently loaded data
and recalculates the sequence logo. This function is only meaningful
for sites known to be palindromic and is especially useful when only
few sequences are available. When searching with palindromised profile,
the "Palindromic" check box should be checked. This function should not
be used with combination with MEME (since MEME itself does a similar
thing).&nbsp; Please also avoid using this function before saving
calibrated profile via the Profile Wisard window, as setting the
"Palindromic" check box in Profile Wisard does exactly the same thing
(an you'll end up wit every sequence duplicated). <br>
&nbsp;&nbsp;&nbsp; All genome search commands are also available from the Genome menu. <br>
&nbsp;&nbsp;&nbsp; The last toolbar button, "<span style="font-weight: bold; color: rgb(0, 102, 0);">Settings</span>" currently allows to set the paths to command line programmes and key scripts used by the GUI. <br>
<br>
<h3><a class="mozTocH3" name="mozTocId23083"></a>Genome Browser Window</h3>
&nbsp;&nbsp; This window is opened after a search for binding sites and could be
used to quickly skim through the sites just found. Alternatively, the
browser could be used to view an existing GenBank file independently of
any search function. The window is split into three viewers which
display feature map (on top), the actual sequence with six frame
translation (in the middle) and search results (in the bottom part).
The feature map is interactive and can be used to select either a
feature by clicking on it or part of the displayed genome fragment by
dragging across it. Selected sequence can be copied to the clipboard,
used as a query to launch database searches, edited or deleted. The
corresponding commands are located in the contextual menu (brought up
by right-clicking in the feature map); you can also double-click a
feature to open its editor. <br>
&nbsp;&nbsp;&nbsp; Database searches can be launched via contextual
menu. Depending on the current selection, the menu will contain
commands to search (with BLAST) against the nr database or (with
hmmer/BLAST) against SwissProt/Uniprot/CDD database. Since NCBI servers
are overloaded most of the time, hmmer searches of SwissProt/UniProt
usually run much faster.&nbsp; <br>
Search results are displayed in the bottom of the window which in
essence is a very simple web browser. Rudimentary navigation here
(Back/Forward/Reload) is possible via the contextual menu.<br>
&nbsp;&nbsp;&nbsp; You can manually resize the top and bottom parts of
this window by dragging the separator (the line with three dots abow
the browser pane) up or down.<br>
&nbsp;&nbsp;&nbsp; The toolbar located on top of this window can be
used to navigate the last hmmer/mast search results (the leftmost
arrows control). These hits could be saved (in fasta format) to a text
file via the corresponding command from the Genome menu. The check box
to the right of this control could be used to exclude the undesired
hits when saving them. The toolbar also allows to zoom in/out feature map (the rightmost
control with +/- signs) or to search within the genome. The "smart"
search field can distinguish three types of queries (sequence,
coordinate or
feature text) and performs the search according to query type.
Navigating to the next search result is posible via the Control-G
shortcat (Command-G on a mac) or the command from the <a href="#mozTocId693367">Genome menu</a>. <br>


<br>
<h3><a class="mozTocH3" name="mozTocId508441"></a>Database Windows</h3>
&nbsp;&nbsp;&nbsp; The <span style="color: rgb(0, 102, 0); font-weight: bold;">RegPrecise and RegulonDB windows</span>
provide access to the corresponding databases with regulatory
informations. These windows have similar organisation and behaviour.
Most of each window is occupied by the regulator list. Since RegulonDB
contains information only for E. coli, the regulators are displayed
straight away, while in case of RegPrecise you have to choose a species
first (from the popup above the list). The top of the RegulonDB window
allows to switch between the transcription and sigma factor&nbsp;
binding sites and filter the sites according to the evidence confidence
level. <br>
&nbsp;&nbsp;&nbsp; Clicking a regulator in the list activates action
buttons in the bottom of the windows. The leftmost button (with an "i"
letter) connects to the database and displays the information on the
regulator in a new window. <br>
&nbsp;&nbsp;&nbsp; The <span style="color: rgb(0, 102, 0); font-weight: bold;">Check TF Presence</span>
button, located to the right of the info button, can be used to verify
the presence of the transcription factor in the currently opened
genome. The button will be disabled if there's no genome opened. This
button connects to the corresponding database to get the amino acid
sequence of the regulator and then launches tfastx search vs the opened
genome. The three topmost hits of this search are displayed in the log
pane of the main window. Since similarity levels vary greatly between
genomes, it's up to the user to estimate its significance. A reciprocal
confirmation could be helpful here: copy the coorginate of the topmost
hit, find the corresponding ORF in the genome browser and launch phmmer
search vs SwissProt/Uniprot to see if the original TF and its obvious
orthologues come up as the top hits.<br>
&nbsp;&nbsp;&nbsp; Please note that the path to the genome file should
not contain spaces! This is due to the way options are treated by
tfastx. <br>
&nbsp;&nbsp;&nbsp; The <span style="color: rgb(0, 102, 0); font-weight: bold;">Regulon Logo</span>
button is located in the lower right of the windows and could be used
to load the binding site data for the currently selected regulator an
display its logo in the main window. The RegPrecise window contains an
additional button to display the logo of the binding sites from the
corresponding regulog. Depending on the number of sites available for
the regulator and their diversity, either button can be more
usable.&nbsp; <br>
<br>
<h3><a class="mozTocH3" name="mozTocId297884"></a>Profile Wizard Window</h3>&nbsp;&nbsp;&nbsp; This window allows to enter the settings required to make a calibrated
profile. The top left of the window contains search thresholds, of
which only the <span style="font-weight: bold; color: rgb(0, 102, 0);">nhmmer gathering threshold</span> is strictly required as it is used by default by nhmmer. Choosing the right threshold can be simplified by the <a href="#FindMinimalScore">Find Minimal Score</a>
command which finds the minimal score for binding sites in the training set.
Although other two nhmmer cutoffs (and mast p-value threshold) are not
required for saving the calibrated .sig file, they are still desirable.
<br>&nbsp;&nbsp;&nbsp; Entering the correct post-processing options in the top right of this
window is critical for making correct additions to genome
annotation.&nbsp; The <span style="font-weight: bold; color: rgb(0, 102, 0);">Palindromic site</span>
check box sets corresponding options when running MEME and MAST and
filters overlapping results produced for palindromic sites by nhmmer.
Checking the <span style="font-weight: bold; color: rgb(0, 102, 0);">Use next locus_tag</span>
option will pick up both the locus_tag and gene qualifiers from the
following gene when possible (these qualifiers won't be added if the
binding site is located between divergently transcribed genes). The <span style="color: rgb(0, 102, 0); font-weight: bold;">Ignore sites within ORFs</span>
option can significantly reduce the amount of non-specific hits for
"noisy" profiles. This option, however, should be used with caution, as
it may also remove some of the specific hits, especially for
repressors. The text entered in the <span style="font-weight: bold; color: rgb(0, 102, 0);">protein name</span> field will be used as the value for the required bound_moiety qualifier when adding sites to genome annotation. <br>&nbsp;&nbsp;&nbsp; Profile description should be entered in the text box in the lower part
of the window. It is required to activate the Save... button and is
expected to include the information on the data source(s) and
description of the profile construction procedure. <br>
&nbsp;&nbsp;&nbsp; If an existing .sig file is opened, all settings
from this file are entered in the fields of the Profile Wizard.
However, these values will be erased if you edit the alignment
sequences. If you want to prevent this, press the <span style="font-weight: bold; color: rgb(0, 102, 0);">Lock</span> button located in the lower right of this window. <br>
<h3><a class="mozTocH3" name="mozTocId65146"></a>Command Configuration Windows</h3>
These simple windows are opened in most cases before launching command
line utilities (nhmmer, meme, mast, TransTerm HP) and python scripts to
allow changing some of the options. The options are hopefully
self-explanatory, but an explanation is provided in most cases via help
tags: hold a mouse over the option for a second to see this help. <br>
<h3><a class="mozTocH3" name="mozTocId686762"></a>Web Browser and Help Windows </h3>
The minimalistic web browser windows are used to display info from
RegPrecise and RegulonDB databases, as well as SigmoID help. If these
seem inconvenient, a link could be copied (via a contextual menu) and
opened in a browser of your choice. <br>
<br>
<h3><a class="mozTocH3" name="mozTocId522291"></a>SigmoID Preferences Window</h3>&nbsp;&nbsp;&nbsp; This window can be opened via the Prefereces... command located in the
SigmoID menu on OS X or in the Edit menu on Windows/Linux or by
pressing the rightmost toolbar button in the main window. The buttons
in the top part of this window switch between three preference panels.
The panels allow to:<br>&nbsp;&nbsp;&nbsp;
1) Set the paths to executable files (nhmmer, meme, mast, etc.) used by
SigmoID. This can be useful if SigmoID can't find some of the required
programmes or if you like to use the ones already installed on your
system. The paths entered here remain valid until SigmoID quits. Once
you've verified that a particular executable file works, please&nbsp;
copy it to the Resources folder within SigmoID folder (on OS X,
right-click SigmoID.app and chose 'Show Package Contents'). <br>&nbsp;&nbsp;&nbsp; 2) Set the databases searched by BLAST and optionally restrict the
searches to a smaller taxonomic group to speed up them. This panel also
allows to switch between two result formats that can be output by the
HMMER web server: the full graphics rich html format (default) and
simple text format. Please note that html format must be supported by
the html engine of your system (which is used by SigmoID), and that
isn't necessarily the case. In particular, no version of Internet
Explorer seems to be capable of displaying the default html HMMER
server output. Neither is WebKit 1.0 that is used by SigmoID on 32-bit
Linux. So, plain text format should be used on Windows and 32-bit Linux.<br>&nbsp;&nbsp;&nbsp; 3) Switch to an alternative folder with calibrated profiles from the
one provided with SigmoID. Only the profiles from this folder will be
accessible via the leftmost toolbar button in the main window. Only
these profiles could be used by the&nbsp; <a href="#Scan_Genome">Scan Genome</a> function.<br>
<br>


<h2><a class="mozTocH2" name="mozTocId385906"></a>Menu Reference<br>
</h2>
<h3><a class="mozTocH3" name="mozTocId324242"></a>File Menu</h3>
&nbsp;&nbsp;&nbsp; This menu contains the standard open and save commands separated into three groups. <br>
&nbsp;&nbsp;&nbsp; The topmost group contains commands related to alignments/profiles, the next one – to genome files.<br>
&nbsp;&nbsp;&nbsp; <br>
&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">Open Profile... </span><br>
&nbsp;&nbsp;&nbsp; Displays an Open File dialog where you can select a
profile/alignment file from your local disk. SigmoID can open files in
its own format (.sig) or text files in fasta format. The file should
have one of the following extensions: .sig, .fasta, .fas, .fsa,
.fa.&nbsp;&nbsp; <br>
<br>
&nbsp;&nbsp;&nbsp; <span style="color: rgb(0, 102, 0); font-weight: bold;">Save Profile As...</span> <br>
&nbsp;&nbsp;&nbsp; becomes enabled if the binding site sequences are
changed. Rather than saving the changes directly, this command opens
the <a href="#mozTocId297884">Profile Wizard</a> window which allows
to enter new profile settings and save the alignment in a .sig file. If
you want to save just the sequences in fasta format, please use the
next command.<br>
<br>
<a name="Save_Profile_Selection"></a><span style="color: rgb(0, 102, 0); font-weight: bold;">Save Profile Selection...</span><br>
<span style="font-weight: normal;">&nbsp;&nbsp;&nbsp; saves (in fasta
format) the part of the profile corresponding to the currently selected
part of the sequence logo.</span><br>

<br>

&nbsp;&nbsp;&nbsp; <span style="color: rgb(0, 102, 0); font-weight: bold;">Save Logo Picture...</span> <br>

&nbsp;&nbsp;&nbsp; does what it says and does it in PNG format. <br>

<br>

&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">Close</span><br>

&nbsp;&nbsp;&nbsp; Closes the current window. The main window can't be closed with this command.<br>

<hr style="width: 100%; height: 2px;">&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">Open Genome...</span><br>
&nbsp;&nbsp;&nbsp; Displays an Open File dialog where you can select a
genome file from your local disk. The file should be in the GenBank
format and have the .gb or .gbk extension.<br>
&nbsp;<br>
&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">Save Genome...</span><br>

&nbsp;&nbsp;&nbsp; Save the file currently opened in the Genome Browser window with the same filename.<br>
<br>
&nbsp;&nbsp;&nbsp; <span style="color: rgb(0, 102, 0); font-weight: bold;">Save Genome As...</span><br>
&nbsp;&nbsp;&nbsp; Saves the genome currently opened in the Genome Browser window with a different filename. <br>
<br>
&nbsp;&nbsp;&nbsp; <span style="color: rgb(0, 102, 0); font-weight: bold;">Export Sequence...</span><br>


&nbsp;&nbsp;&nbsp; Export the contents of the current genome as plain text file in fasta format. This discards feature table.<br>
<br>
&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">Export Feature Table...</span><br>
&nbsp;&nbsp;&nbsp; Export feature table in GenBank Sequin table format.
The resulting .tbl file can be used to prepare GenBank submission with
the help of <a href="http://www.ncbi.nlm.nih.gov/genbank/tbl2asn2/">tbl2asn</a>.<br>
<hr style="width: 100%; height: 2px;">&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">Quit</span><br>

&nbsp;&nbsp;&nbsp; Closes all SigmoID windows and exits SigmoID completely. If you
select this option with unsaved profile or genome, SigmoID will first
ask you to save the changes.<br>
<br>
<br>
<h3><a class="mozTocH3" name="mozTocId121686"></a>Edit Menu</h3>
<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; Undo</span><br>
&nbsp;&nbsp;&nbsp; Undoes the last editing action done in the currently
active text field. Unfortunately, Undo is not available for changes
made to genome files.<br>
<hr style="width: 100%; height: 2px;"><span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; Cut</span><br>

&nbsp;&nbsp;&nbsp; Copies the selected text to the clipboard and deletes it from the original position.<br>
<br>
&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">Copy</span><br>
&nbsp;&nbsp;&nbsp; Copies the selected text to the clipboard. In Genome
Browser window this command copies the nucleotide sequence. <br>
<br>
&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">Copy Protein Sequence</span><br>
&nbsp;&nbsp;&nbsp; If a CDS is currently selected in the Genome
Browser, this command copies its amino acid sequence to the clipboard.<br>
&nbsp; <br>
&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">Paste</span><br>
&nbsp;&nbsp;&nbsp; Pastes the text from the clipboard copied using Cut or Copy command to the current cursor location.<br>
<br>
<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; Clear</span><br>


&nbsp;&nbsp;&nbsp; Deletes the selected text.<br>

<hr style="width: 100%; height: 2px;">&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">Select All</span><br>

&nbsp;&nbsp;&nbsp; Selects all text in the current text field<br>
<hr style="width: 100%; height: 2px;"><span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; Preferences</span><br>
&nbsp;&nbsp;&nbsp; Opens the Preferences window to change personal
preferences for SigmoID. Currently only alows to set the paths to
executable files. On OS X this submenu is located in the SigmoID menu. <br>
<br>
<br>
<h3><a class="mozTocH3" name="mozTocId916489"></a>View Menu</h3>
This menu can be used to change the information displayed in the <a href="#mozTocId965351">Main window</a> and in the <a href="#mozTocId23083">Genome Browser</a>. Only the last command, <span style="font-weight: bold; color: rgb(0, 102, 0);">View Details</span>, is related to the <a href="#mozTocId23083">Genome Browser</a>. The remaining commands are related to the main window and swich the type of information displayed in the topmost <a href="#Viewer">Viewer</a>
pane. This menu allows to view contents of all components of a .sig
file which is actually a virtual folder containing six text files. This
meny allows to view the information contained within the .sig files.
Editing the sequences can be done directly in the Viewer, while the
rest of information can&nbsp; only be edited via the <a href="#mozTocId297884">Profile Wizard</a>.  <br>
<br>
&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">Logo<br>
</span><span style="color: black;"></span>&nbsp;&nbsp;&nbsp; Displays
sequence logo for the sequences in the currently opened .sig or fasta
file (or downloaded from the RegPrecise or RegulonDB databases).
Currently, the logo is calculated using the original T. Schneider
(1986) formula without small sample correction.<br>
<br>
<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; Sequences</span><br>&nbsp;&nbsp;&nbsp;
Shows the actual nucleotide sequences for the loaded binding site data.
You can edit the sequences in this view, but if you wish to use the
edited data further, please switch to the Logo view, as this finalises
your editing and recalculates the logo and the hmm profile.<br>
<br>
&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">Profile Info</span><br>


&nbsp;&nbsp;&nbsp; The description of the profile as given by its author. Available only for data from .sig files. <br>
<br>

<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp;&nbsp;</span><span style="font-weight: bold; color: rgb(0, 102, 0);">Hmm Profile</span><br>&nbsp;&nbsp;&nbsp;
The calibrated hmm profile produced by hmmbuild when creating the .sig
file. Available only for data from .sig files.<br>
<br>

<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; MEME data</span><span style="font-weight: bold; color: rgb(0, 102, 0);"></span><br>


&nbsp;&nbsp;&nbsp; The same sequences in MEME format. These are used for MAST searches. Available only for data from .sig files.<br>
<br>

<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; </span> <span style="font-weight: bold; color: rgb(0, 102, 0);">Profile Settings</span><br>&nbsp;&nbsp;&nbsp;
Various settings, including profile calibration thresholds and post
processing options. Available only for data from .sig files.<span style="font-weight: bold; color: rgb(0, 102, 0);">
 </span>
<hr style="width: 100%; height: 2px;">&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">Hide Viewer<br>
&nbsp;&nbsp;&nbsp; </span>hiddes/unhides via the  <a href="index.html#Viewer">Viewer</a> in the main window to give less/more room for the Log pane.<br>
<hr style="width: 100%; height: 2px;">&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">View Details<br>
&nbsp;&nbsp;&nbsp; </span>shows/hides sequence display with six frame translation in the Genome Browser window.<br>
<br>
<br>
<h3><a class="mozTocH3" name="mozTocId494918"></a>Profile Menu</h3>
&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">Extend Binding Sites...</span><br>
&nbsp;&nbsp;&nbsp; Opens a small window where you can specify the left
and right extension limits, as well as the genome file to search. This
command finds every sequence from the currently opened binding site
data in the genome sequence and adds the specified number of bases to
the left and to the right. The results are written to the log pane. <br>
<hr style="width: 100%; height: 2px;"><span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; Convert to Stockholm</span><br>
&nbsp;&nbsp;&nbsp; Converts the current profile to the minimal
Stockholm format (as required by hmmbuild) and outputs the results to
the <a href="#LogPane">Log</a> pane.<br>
<br>

&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">Convert to Hmm</span><br>



&nbsp;&nbsp;&nbsp; Runs hmmbuild from the HMMER package to create a hmm profile that could be used as input for nhmmer.<br>
<br>


<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; </span><span style="font-weight: bold; color: rgb(0, 102, 0);">Convert to MEME</span><br>



&nbsp;&nbsp;&nbsp; Runs MEME with the currently loaded binding site sequences and outputs the results as plain text to the <a href="index.html#LogPane">Log</a>
pane. These results san be used as input by MAST (in fact, when running
MAST with uncalibrated data MEME is run first in exactly the same way)<br>

<hr style="width: 100%; height: 2px;"><span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; <a name="Find_Sites_with_MEME"></a>Find Sites with MEME...</span><span style="font-weight: bold; color: rgb(0, 102, 0);"></span><br>&nbsp;&nbsp;&nbsp;
Shows a window allowing you to configure MEME parameters. For this
command MEME is configured to produce results in html format, hence
they are displayed in the <a href="#mozTocId686762">Web Browser</a> window. This command may be useful when dealing with unalugned data, e.g. from RegulonDB.
<hr style="width: 100%; height: 2px;">&nbsp;&nbsp;&nbsp;&nbsp;  <span style="font-weight: bold; color: rgb(0, 102, 0);"></span><span style="font-weight: bold; color: rgb(0, 102, 0);">Profile Wizard...</span><br>
&nbsp;&nbsp;&nbsp;&nbsp; Opens the <a href="#mozTocId297884">Profile Wizard</a> window which allows to enter calibrated profile settings and then save it in as a .sig file. <br>
<br>
<br>
<h3><a class="mozTocH3" name="mozTocId972700"></a>Regulon menu</h3>&nbsp;&nbsp;&nbsp; The first two items of this menu provide access to databases with regulon information.  The <span style="font-weight: bold; color: rgb(0, 102, 0);">RegPrecise</span> database contains high quality information on binding sites for many bacteria while the <span style="font-weight: bold; color: rgb(0, 102, 0);">RegulonDB</span> is a specialised <span style="font-style: italic;">E. coli</span>
regulon database. While the information from RegulonDB in most cases
requires <a href="#Find_Sites_with_MEME">additional steps</a> to be usable, it has data on
regulators not present in RegPrecise. When using the data from RegulonDB for genomes other than <span style="font-style: italic;">E. coli</span>, it might be worthwhile to check for the presence of the TF orthologue in the studied genome. This can be done with the <a style="color: rgb(0, 0, 238);" href="#Check_TF_Presence">Check TF Presence</a>
command.<br>
<br>
&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">RegPrecise...</span><span style="font-weight: bold; color: rgb(0, 102, 0);"></span><br>


&nbsp;&nbsp;&nbsp; Opens the <a href="#mozTocId508441">window</a> providing access to the <a href="http://Regprecise.lbl.gov/">RegPrecise database</a>.<br>
<br>

<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; </span><span style="font-weight: bold; color: rgb(0, 102, 0);">RegulonDB...</span><br>


&nbsp;&nbsp;&nbsp; Opens the <a href="index.html#mozTocId508441">window</a> providing access to the <a href="http://regulondb.ccg.unam.mx">RegulonDB database</a>.<br>


<hr style="width: 100%; height: 2px;">&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">Regulon Info</span><br>&nbsp;&nbsp;&nbsp; Opens the RegPrecise or RegulonDB web page with info for the regulon currently selected in one of the <a href="#mozTocId508441">database windows</a>.<br>
<br>


<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; Show Logo</span><span style="font-weight: bold; color: rgb(0, 102, 0);"></span><br>&nbsp;&nbsp;&nbsp; Shows the logo of the currently selected regulator
binding sites in the main SigmoID window. The result can usually be
used for nhmmer/mast search for RegPrecise data. The binding sites in
RegulonDB are often misaligned, in which case the <span style="font-weight: bold; color: rgb(0, 102, 0);">Find Sites with MEME</span> command from the <span style="color: rgb(0, 102, 0); font-weight: bold;">Profile</span> menu may (or may not) be useful. &nbsp;&nbsp;  <br>
<br>


<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; <a name="Check_TF_Presence"></a>Check TF Presence</span><span style="font-weight: bold; color: rgb(0, 102, 0);"></span><br>



&nbsp;&nbsp;&nbsp; This command gets the regulator protein sequence from
RegulonDB and runs tfastx search versus the currently open genome. The
top three tfastx search results are displayed in the log pane of the
main SigmoID window. We recommend a reciprocal check using the
coordinates of the best hit to locate it in the current genome and run
the phmmer search vs the SwissProt database which (in case of
orthology) should bring the original regulator as the best <span style="font-style: italic;">E. coli </span>hit. At the moment this
command is not available for RegPrecise since there's no
straightforward way to get regulator sequences from this database.
<hr style="width: 100%; height: 2px;"><span style="font-weight: bold; color: rgb(0, 102, 0);"></span><span style="color: rgb(0, 102, 0); font-weight: bold;"><a name="FindMinimalScore"></a>Find Minimal Score</span>
<br>
&nbsp;&nbsp;&nbsp; May be helpful when determining search thresholds. This command
can only be issued when a nhmmer search has just been run and its
results are displayed in the Genome Browser window. This function
simply compares the current hits to the training set (original binding
sites opened in the main window), outputs the lowest score among the
training set and lists missed hits. The lowest score found is also entered as the nhmmer gathering threshold into the <a href="#mozTocId297884">Profile Wizard</a> window.<br>




&nbsp;&nbsp;&nbsp; <br>
<br>
<br>
<h3><a class="mozTocH3" name="mozTocId693367"></a>Genome Menu</h3>&nbsp;&nbsp;&nbsp;
This menu collects genome related commands and is mostly oriented at
various ways of searching for regulatory information in currently
opened genome. <br>

<br>
&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">nhmmer Search...<br>
</span>&nbsp;&nbsp;&nbsp; Opens the window allowing to configure and
launch nhmmer, which is the primary search engine in SigmoID. This
function is enabled if binding site data is loaded (and sequence logo
of the site is displayed in the main window).&nbsp; To launch the
search, you have to choose a file with genome sequence in GenBank
format and choose the cutoff score (which is critical for getting the
correct results). If a calibrated profile is loaded, then the correct
cutoff will be chosen already. The raw search results appear in the <span style="font-weight: bold; color: rgb(0, 102, 0);">Log</span>
pane. If you are sure that the cutoff is correct, you may check the
"Add annotation to the genome" check box which will run the HmmGen.py
python script to filter nhmmer results, add the binding sites to genome
annotation and open the updated genome sequence in the <a href="index.html#mozTocId23083">Genome Browser window</a>. In case a calibrated profile (.sig file) is opened, this script is launched&nbsp; by default. For non-calibrated profiles
(opened from fasta files) the post processing script has to be launched
separately by the <a href="#Annotate_Current_Sites">Annotate Current Sites</a> command. <br>
<br>


&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);"><a name="Annotate_Current_Sites"></a>Annotate Current Sites...</span><br>

&nbsp;&nbsp;&nbsp; Opens the window allowing to configure and launch python script
(<a href="#HmmGen.py">HmmGen.py</a>) to filter nhmmer results, add the binding sites to genome
annotation and open the updated genome sequence in the <a href="index.html#mozTocId23083">Genome Browser window</a>.
Using this command separately from nhmmer search may be convenient for
unoptimised profiles when desiding on the correct search options.<br>
<br>

<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; MAST </span><span style="font-weight: bold; color: rgb(0, 102, 0);">Search...</span><br>&nbsp;&nbsp;&nbsp;
Opens the window allowing to launch MAST from the MEME Suite package.
Please note that compared to nhmmer MAST wasn't extensively tested
within SigmoID. The configuration window provides minimum options
(basically, just one – the p-value cutoff), but allows to enter
additional options if desired. These will be appended to the end of
MAST command line. If a non-calibrated profile is opened, MEME is run
before MAST to convert the binding site sequences to the required
format. The raw search
results (and, in case MEME was run, its output as well)&nbsp; appear in
the <span style="font-weight: bold; color: rgb(0, 102, 0);">Log</span>
pane. The checkboxes in the bottom of this window instruct SigmoID to show the results filtered by the post processing script (<a href="#MastGen.py">MastGen.py</a>) in the <a href="index.html#mozTocId23083">Genome Browser window</a> with or without modifying the annotation.<br>
<br>


<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; </span><span style="font-weight: bold; color: rgb(0, 102, 0);">Terminator </span><span style="font-weight: bold; color: rgb(0, 102, 0);">Search...</span><br>&nbsp;&nbsp;&nbsp;
Opens the window allowing to configure and launch TransTerm HP to
search for terminators. This command performs the necessary format
conversions and adds the
terminators to genome annotation (with the <a href="#TermGen.py">TermGen.py</a> script). As TransTerm takes some time to run,
the results may take couple of minutes to appear, depending on
available processing power. <br>
<br>
&nbsp;&nbsp;&nbsp; <a name="Scan_Genome"></a><span style="font-weight: bold; color: rgb(0, 102, 0);">Scan Genome...<br>
&nbsp;&nbsp;&nbsp; </span><span style="color: black;">This command is
designed to perform a full genome scan with all available calibrated
profiles with minimal interaction with the user. You can select the
desired profiles (or use all of them) and choose if terminator search
will be performed. After pressing the Run button, SigmoID will run
nhmmer followed by the HmmGen.py script with all checked profiles using
preconfigured settings followed by the terminator search. The results
will be written to the file specified by you in the GenBank format.
With many profiles this function may take a while to run.</span><span style="font-weight: bold; color: rgb(0, 102, 0);"><br>
</span>
<hr style="width: 100%; height: 2px;"><span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; Save Checked Sites...</span><span style="font-weight: bold; color: rgb(0, 102, 0);"></span><br>&nbsp;&nbsp;&nbsp; Saves to a text
file (in fasta format) the hits from the last search currently displayed in the Genome Browser. The check box
to the right of the navigation arrows in the toolbar of Genome Browser can be used to exclude the undesired
hits when saving them.&nbsp;&nbsp;  <br>
<br>


<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; List Regulons...</span><span style="font-weight: bold; color: rgb(0, 102, 0);"></span><br>&nbsp;&nbsp;&nbsp;
Outputs to the Log pane of the main window either a single regulon
(controlled by the specified regulator) or all regulons currently
annotated in the genome. The regulons are output by sequentially
listing the operons/divergons controlled by a regulator. For the
purpose of this command the operon is defined as the genes between a
binding site and the nearest terminator or a long intergenic gap. Two
divergently transcribed operons are output as a single divergon with
the regulator site in the middle. The settings window opened by this
command allows to choose the criteria for operon beginning and start.<br>
<hr style="width: 100%; height: 2px;">&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);"></span><span style="font-weight: bold; color: rgb(0, 102, 0);">Find</span><br>
&nbsp;&nbsp;&nbsp; This command simply puts the cursor into the search
field of the Genome Browser window located in its top right corner. The
search field can distinguish three types of queries (sequence,
coordinate or
feature text) and performs the search according to the type of query.
Type what to search for and press "Enter" on the keyboard to initiate
the search.
The position of the first occurrence of the query in the genome will be
highlighted.<br>
<br>

<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; </span><span style="font-weight: bold; color: rgb(0, 102, 0);">Find Again</span><br>


&nbsp;&nbsp;&nbsp; Highlights the next position in genome of the previous search query.<br>
<span style="font-weight: bold; color: rgb(0, 102, 0);"></span><br>



&nbsp;&nbsp;&nbsp; <br>


<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; </span><br>
 
<h3><a class="mozTocH3" name="mozTocId600980"></a>
Window Menu</h3>

&nbsp;&nbsp;&nbsp; Lists all currently opened SigmoID windows. Selecting a window from this list brings it to the front.<br>
<br>
<br>
<br>
<h3><a class="mozTocH3" name="mozTocId2010"></a>
Help Menu</h3>

&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">About SigmoID</span><br>


&nbsp;&nbsp;&nbsp; Displays a window with information about SigmoID including the current version and a brief list of credits.<br>
<br>

<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; SigmoID Help</span><br>


&nbsp;&nbsp;&nbsp; Opens the Help Viewer window.
<hr style="width: 100%; height: 2px;">
&nbsp;&nbsp;&nbsp; <span style="font-weight: bold; color: rgb(0, 102, 0);">HMMER User Guide</span><br>



&nbsp;&nbsp;&nbsp; Opens (in the default PDF viewer) the HMMER User guide from the HMMER package distribution.<br>
<br>


<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp;&nbsp;</span><span style="font-weight: bold; color: rgb(0, 102, 0);">Hmmer.org</span><br>



&nbsp;&nbsp;&nbsp; Opens HMMER website in the web browser window.<br>
<hr style="width: 100%; height: 2px;">

<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; MEME Suite Web Portal</span><span style="font-weight: bold; color: rgb(0, 102, 0);"></span><br>



&nbsp;&nbsp;&nbsp; Opens the main MEME Suite website in the web browser window.<span style="font-weight: bold; color: rgb(0, 102, 0);">&nbsp;&nbsp;&nbsp; <br>
</span><p style="margin-left: 1.03cm;"><br>
</p>
<h1><a class="mozTocH1" name="mozTocId708060"></a><a name="Python_Scripts"></a>Python Scripts</h1>
&nbsp;&nbsp;&nbsp; The scripts described below process output produced
by various search programmes, perform format conversions and add
features to genome annotation. The scripts are called by SigmoID GUI
when necessary, but can be used separately if desired. Type the command below in terminal to get help on command line usage:<br>
<span style="font-family: monospace;">&nbsp;&nbsp; <br>
&nbsp;&nbsp; python &lt;path_to_the_script&gt; -h</span><br style="font-family: monospace;">
<br>
<h2><a class="mozTocH2" name="mozTocId227412"></a><a name="HmmGen.py"></a>HmmGen.py</h2>&nbsp;&nbsp;&nbsp; SigmoID processes nhmmer results (table of hits) with the help of the <span style="font-weight: bold; color: rgb(0, 102, 0);">HmmGen.py</span>
python script, adding corresponding feature annotations to the genbank
file being searched and saving the result in a new file. Some useful
options are provided to make annotation more convenient. These you can
find in the "<span style="font-weight: bold; color: rgb(0, 102, 0);">HmmGen Settings</span>" window, which pops up after clicking the "<span style="font-weight: bold; color: rgb(0, 102, 0);">Postprocess</span>" button in the main window. <br>
&nbsp;&nbsp;&nbsp; To run the script, enter the appropriate threshold
(either bit score or E-value). By default SigmoID chooses the same
value that was used to run <span style="font-weight: bold;">nhmmer</span>, but you can increase the bit score or decrease the E-value to reduce the number of hits without re-running nhmmer. <br>
&nbsp;&nbsp;&nbsp; To filter out all intragenic hits, check the "<span style="font-weight: bold;">Consider intergenic regions only</span>" box. <span style="font-weight: bold;">nhmmer</span>
reports hits on both strands, and in the case of palindromic sites
there will be two hits with the same coordinates and identical (or very
close) scores. To remove one of the duplicate sites, check the "<span style="font-weight: bold;">Palindromic site</span>" checkbox. <br>
&nbsp;&nbsp;&nbsp; This script can also add 'locus_tag' and 'gene'
qualifiers to the feature being annotated, but please note that GenBank
will object such additions if you later decide to submit this sequence
to the database. If you are certain you really want this addition,
check the "<span style="font-weight: bold;">Add qualifier</span>" box.<br>
&nbsp;&nbsp;&nbsp; Choose feature type ("promoter" or "protein_bind") from the "<span style="font-weight: bold;">Feature to add:</span>"
box (or just type in the valid feature type). The window also allows to
configure one qualifier for this feature. The qualifier name could be
typed in, but it should remain as 'protein_bind' in most cases. A valid
protein name should be entered in the field to the right.<br>
&nbsp;&nbsp;&nbsp; Pressing the <span style="font-weight: bold; color: rgb(0, 102, 0);">Run</span>
button will ask you for the name of the file in which you'd like to
save the genome sequence with modified annotation. If the "Show hits in
genome browser" box is checked, You'll see the results in the browser
window. The script also appends the detailed text report to the log
pane.<br>

<br>
<h2><a class="mozTocH2" name="mozTocId61963"></a><a name="MastGen.py"></a>MastGen.py</h2>
&nbsp;&nbsp;&nbsp; This script allows to add features to a genbank file
according to MAST results. From SigmoID it is called when <br>
<br>
<pre>usage: </pre>
<pre>MastGen &lt;report_file&gt;&nbsp; &lt;input_file&gt; &lt;output_file&gt; [options]</pre>
<pre><br>positional arguments:</pre>
<pre>&nbsp; report_file&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; path to MAST report file produced with -tblout option.</pre>
<pre>&nbsp; input_file&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; path to input GenBank file.</pre>
<pre>&nbsp; output_file&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; path to output GenBank file.</pre>
<pre><br>optional arguments:</pre>
<pre>&nbsp; -h, --help&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; show this help message and exit</pre>
<pre>&nbsp; -L &lt;integer&gt;, --length &lt;integer&gt;</pre>
<pre>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; final feature's length in genbank file</pre>
<pre>&nbsp; -q [&lt;key#"value"&gt; [&lt;key#"value"&gt; ...]], --qual [&lt;key#"value"&gt; [&lt;key#"value"&gt; ...]]</pre>
<pre>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; add this qualifier to each annotated feature.</pre>
<pre>&nbsp; -p, --palindromic&nbsp;&nbsp;&nbsp;&nbsp; filter palindromic sites.</pre>
<pre>&nbsp; -n, --name&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; don't pick 'locus_tag' and 'gene' qualifiers from the</pre>
<pre>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; next CDS feature.</pre>
<pre>&nbsp; -V &lt;float or integer&gt;, --pval &lt;float or integer&gt;</pre>
<pre>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; threshold E-Value.</pre>
<pre>&nbsp; -S &lt;float or integer&gt;, --score &lt;float or integer&gt;</pre>
<pre>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; threshold Bit Score.</pre>
<pre>&nbsp; -i, --insert&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; don't add features inside CDS</pre>
<pre>&nbsp; -d, --duplicate&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; no duplicate features with the same location and the</pre>
<pre>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; same protein_bind qualifier value</pre>
<pre>&nbsp; -v, --version&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; show program's version number and exit</pre>
<pre>&nbsp; -f &lt;"feature key"&gt;, --feature &lt;"feature key"&gt;</pre>
<pre>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; feature key to add (promoter, protein_bind etc.)</pre>
<br>
<h2><a class="mozTocH2" name="mozTocId891699"></a><a name="TermGen.py"></a>TermGen.py</h2>
&nbsp;&nbsp;&nbsp; This script allows to add terminators to a genbank file according to TransTerm HP results.<br>
<br>
<pre>usage: </pre>
<pre>TermGen &lt;input_file&gt; &lt;output_file&gt; [options]</pre>
<pre><br>positional arguments:</pre>
<pre>&nbsp; input_file&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; path to input GenBank file.</pre>
<pre>&nbsp; output_file&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; path to output GenBank file.</pre>
<pre><br>optional arguments:</pre>
<pre>&nbsp; -h, --help&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; show this help message and exit</pre>
<pre>&nbsp; -o &lt;path&gt;, --output &lt;path&gt;</pre>
<pre>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; redirects TransTerm HP output file to directory given</pre>
<pre>&nbsp; -C &lt;integer&gt;, --confidence &lt;integer&gt;</pre>
<pre>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; threshold Score.</pre>
<pre>&nbsp; --minstem &lt;integer&gt;&nbsp;&nbsp; Stem must be n nucleotides long</pre>
<pre>&nbsp; --minloop &lt;integer&gt;&nbsp;&nbsp; Loop portion of the hairpin must be at least n long</pre>
<pre>&nbsp; --maxlen &lt;integer&gt;&nbsp;&nbsp;&nbsp; Total extent of hairpin &lt;= n NT long</pre>
<pre>&nbsp; --maxloop &lt;integer&gt;&nbsp;&nbsp; The loop portion can be no longer than n</pre>
<pre>&nbsp; -v, --version&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; show program's version number and exit</pre>
<br>
<h2><a class="mozTocH2" name="mozTocId793586"></a>ptt_converter.py</h2>
This script allows to convert genbank file into .ptt file format.<br>
<br>
<pre>usage: </pre>
<pre>Genbank to PTT converter &lt;input_file&gt; </pre>
<pre><br>positional arguments:</pre>
<pre>&nbsp; input_file&nbsp;&nbsp;&nbsp;&nbsp; path to input Genbank file.</pre>
<pre><br>optional arguments:</pre>
<pre>&nbsp; -h, --help&nbsp;&nbsp;&nbsp;&nbsp; show this help message and exit</pre>
<pre>&nbsp; -v, --version&nbsp; show program's version number and exit<br><br></pre>
<h2><a class="mozTocH2" name="mozTocId218330"></a>OperOn.py</h2>
&nbsp;&nbsp;&nbsp; This script finds putative operons between regulator binding sites and/or terminators/long intergenic gaps. <br>
<br>
<pre>usage: </pre>
<pre>OperOn &lt;input_file&gt; [options]</pre>
<pre><br>positional arguments:</pre>
<pre>&nbsp; input_file&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; path to input GenBank file.</pre>
<pre><br>optional arguments:</pre>
<pre>&nbsp; -h, --help&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; show this help message and exit</pre>
<pre>&nbsp; -g &lt;int&gt;, --gap &lt;int&gt;</pre>
<pre>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; minimal gap between operons</pre>
<pre>&nbsp; -i &lt;int&gt;, --indent &lt;int&gt;</pre>
<pre>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; maximal distance from binding site to the first</pre>
<pre>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; downstream CDS</pre>
<pre>&nbsp; -t, --terminator&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; terminators are regarded as operon separator</pre>
<pre>&nbsp; -r &lt;name of regulator&gt;, --regulator &lt;name of regulator&gt;</pre>
<pre>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; only specified regulators are considered</pre>
<pre>&nbsp; -p, --palindromic&nbsp;&nbsp;&nbsp;&nbsp; treat all binding sites as palindromic</pre>
<pre>&nbsp; -s, --strict&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; operon stops on first terminator (if -t is set)</pre>
<pre>&nbsp; -v, --version&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; show program's version number and exit</pre>
<h2><a class="mozTocH2" name="mozTocId385189"></a>gbk2tbl.py</h2>
This script allows to convert GenBank file into .tbl file format. The resulting table is output to stdout.<br>
<br>
<pre>usage: </pre>
<pre>Genbank to .tbl converter &lt;input_file&gt; [options]</pre>
<pre><br>positional arguments:</pre>
<pre>&nbsp; input_file&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; path to input GenBank file.</pre>
<pre><br>optional arguments:</pre>
<pre>&nbsp; -h, --help&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; show this help message and exit</pre>
<pre>&nbsp; -f, --fasta&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; creates fasta from genbank file.</pre>
<pre>&nbsp; -p PREFIX, --prefix PREFIX</pre>
<pre>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; sequencing centre prefix.</pre>
<pre>&nbsp; -t, --translation&nbsp;&nbsp;&nbsp;&nbsp; adds translation qualifier to CDS features in .tbl</pre>
<pre>&nbsp; -v, --version&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; show program's version number and exit</pre>
<br>
<br>
<p style="margin-left: 1.03cm;"><br>
</p>
<p style="margin-left: 1.03cm;"><br>
</p>

</body></html>